File: 01-Programming-Languages/PHP/Algorithms/Graph/bfs.php
<?php

/*
Problem: Breadth-First Search (BFS)
Description: Traverse a graph level by level using a queue.

Approach:
- Use a queue to process nodes FIFO
- Track visited nodes to avoid cycles
- Process all neighbors before moving to next level

Time Complexity: O(V + E)
Space Complexity: O(V)

Example:
Input:  adjacency list: 0->[1,2], 1->[0,3,4], 2->[0], 3->[1], 4->[1], start = 0
Output: 0 1 2 3 4
*/

function bfs(array $graph, int $start): array {
    $visited = array_fill(0, count($graph), false);
    $queue = [$start];
    $visited[$start] = true;
    $result = [];
    while (!empty($queue)) {
        $node = array_shift($queue);
        $result[] = $node;
        foreach ($graph[$node] as $neighbor) {
            if (!$visited[$neighbor]) {
                $visited[$neighbor] = true;
                $queue[] = $neighbor;
            }
        }
    }
    return $result;
}

$graph = [
    [1, 2],
    [0, 3, 4],
    [0],
    [1],
    [1],
];

echo "BFS starting from 0: " . implode(" ", bfs($graph, 0)) . "\n";

File: 01-Programming-Languages/C++/Algorithms/Graph/bfs.cpp
/*
Problem: BFS on Graph (Adjacency Matrix)
Traverse a graph using Breadth-First Search starting from a given source vertex.

Approach:
- Use a queue, mark visited nodes, process level by level

Time Complexity: O(V^2)
Space Complexity: O(V)

Example:
Input: 5 vertices, edges: (0,1),(0,2),(1,3),(1,4), start = 0
Output: 0 1 2 3 4
*/

#include <iostream>
#include <queue>
#include <vector>
using namespace std;

void bfs(const vector<vector<int>>& graph, int start)
{
    int n = graph.size();
    vector<bool> visited(n, false);
    queue<int> q;
    visited[start] = true;
    q.push(start);
    while (!q.empty())
    {
        int u = q.front(); q.pop();
        cout << u << " ";
        for (int v = 0; v < n; v++)
            if (graph[u][v] && !visited[v])
            {
                visited[v] = true;
                q.push(v);
            }
    }
    cout << endl;
}

int main()
{
    vector<vector<int>> graph = {
        {0, 1, 1, 0, 0},
        {1, 0, 0, 1, 1},
        {1, 0, 0, 0, 0},
        {0, 1, 0, 0, 0},
        {0, 1, 0, 0, 0}
    };
    cout << "BFS from vertex 0: ";
    bfs(graph, 0);
    return 0;
}
